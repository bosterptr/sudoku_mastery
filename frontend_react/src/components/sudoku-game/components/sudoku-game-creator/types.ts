// eslint-disable-next-line no-shadow
export const enum StepType {
  SetNumber = 0,
  SetNote = 1,
}
export const SudokuStageOfTheGame = {
  INITIAL: 0,
  WON: 1,
  LOST: 2,
} as const;
export type ISudokuStageOfTheGame =
  (typeof SudokuStageOfTheGame)[keyof typeof SudokuStageOfTheGame];
export type ISudokuValues = string;
// 0 index based
export type INotes = number[];
/** e.g. '3.8.15...4.97....5561..8...9.4...53..5...1.49......8...3...7..67..39.2......8.7..' */
export type IMission = string;
/** XXY = XX-cell index, Y-value. e.g., '205' represents cell index 20 and value 5. */
export type IScoreHistory = string[];
export interface IStep {
  index: number;
  previous: string;
  type: StepType;
  value: string;
}
export type ICellNotes = number[];
