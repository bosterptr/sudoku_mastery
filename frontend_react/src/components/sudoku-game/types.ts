// eslint-disable-next-line no-shadow
export const enum StepType {
  SetNumber = 0,
  SetNote = 1,
}
export const SudokuStageOfTheGame = {
  INITIAL: 0,
  WON: 1,
  LOST: 2,
} as const;
export type ISudokuStageOfTheGame =
  (typeof SudokuStageOfTheGame)[keyof typeof SudokuStageOfTheGame];
export type ISudokuValues = string;
// 0 index based
export type INotes = number[];
// eslint-disable-next-line no-shadow
export const SudokuDifficulty = {
  Easy: 'easy',
  Medium: 'normal',
  Hard: 'hard',
  Expert: 'extreme',
} as const;
export interface ISudoku {
  id: string;
  body: string;
  difficulty: (typeof SudokuDifficulty)[keyof typeof SudokuDifficulty];
  created_at: string;
  updated_at: string;
  user_id: string;
}
/** e.g. '3.8.15...4.97....5561..8...9.4...53..5...1.49......8...3...7..67..39.2......8.7..' */
export type IMission = string;
/** XXY = XX-cell index, Y-value. e.g., '205' represents cell index 20 and value 5. */
export type IScoreHistory = string[];
export interface IStep {
  index: number;
  previous: string;
  type: StepType;
  value: string;
}
export type ICellNotes = number[];
