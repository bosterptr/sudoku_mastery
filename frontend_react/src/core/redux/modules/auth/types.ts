export interface SessionUser {
  id: string;
  email: string;
  displayName: string;
}

export interface APIState {
  done: boolean;
  hasErrored: boolean;
  isLoading: boolean;
}
// eslint-disable-next-line no-shadow
export const AuthAPIStateKeys = {
  loginWithEmail: 0,
  changePassword: 1,
  changePasswordWithToken: 2,
  activateAccount: 3,
  forgotPassword: 4,
} as const;
export type IAuthAPIStateKeys = (typeof AuthAPIStateKeys)[keyof typeof AuthAPIStateKeys];
export interface AuthAPIState {
  [AuthAPIStateKeys.loginWithEmail]?: APIState;
  [AuthAPIStateKeys.changePassword]?: APIState;
  [AuthAPIStateKeys.changePasswordWithToken]?: APIState;
  [AuthAPIStateKeys.activateAccount]?: APIState;
  [AuthAPIStateKeys.forgotPassword]?: APIState;
}
export type IAuthState = AuthAPIState;
