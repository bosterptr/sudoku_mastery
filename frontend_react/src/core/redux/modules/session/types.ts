export interface SessionUser {
  id: string;
  email: string;
  displayName: string;
}

export interface ISessionState {
  firstPageLoaded: boolean;
  user: SessionUser | null;
}
