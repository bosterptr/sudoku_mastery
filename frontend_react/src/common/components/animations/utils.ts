import { CSSProperties } from 'react';

export const reflow = (node: Element) => node.scrollTop;

interface ComponentProps {
  easing: string | { enter?: string; exit?: string } | undefined;
  style: CSSProperties | undefined;
  timeout: number | { enter?: number; exit?: number };
}

interface Options {
  mode: 'enter' | 'exit';
}

interface TransitionProps {
  duration: string | number;
  easing: string | undefined;
  delay: string | undefined;
}

export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
  const { timeout, easing, style = {} } = props;

  return {
    duration:
      style.transitionDuration ??
      (typeof timeout === 'number' ? timeout : timeout[options.mode] || 0),
    easing:
      style.transitionTimingFunction ??
      (typeof easing === 'object' ? easing[options.mode] : easing),
    delay: style.transitionDelay,
  };
}

export function getAutoHeightDuration(height: number) {
  if (!height) return 0;
  const constant = height / 36;
  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
