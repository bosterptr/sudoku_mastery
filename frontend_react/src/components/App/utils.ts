const GL_BUFFER_STATUS_CHECK_INTERVAL = 2;

export type IOpenglSetup = [
  OffscreenCanvas,
  WebGLProgram,
  WebGLProgram,
  WebGLProgram,
  ITexFbPair,
  ITexFbPair,
  number,
];

export interface ITexFbPair {
  tex: WebGLTexture;
  fb: WebGLFramebuffer;
}
export const getGrayScaleFromUint8Array = (
  dataArray: Uint8Array,
  width: number,
  height: number,
  cachedGrayScaleUintArray: Uint8Array,
  cachedGrayScaleSize: number,
): Uint8Array => {
  if (cachedGrayScaleSize !== width * height) {
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray = new Uint8Array(width * height);
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleSize = width * height;
  }
  const { length } = dataArray;
  for (let i = 0, luminanceIndex = 0; i < length; i += 4, luminanceIndex += 1) {
    // eslint-disable-next-line no-bitwise
    const r = dataArray[i] >> 2;
    // eslint-disable-next-line no-bitwise
    const g = dataArray[i + 1] >> 1;
    // eslint-disable-next-line no-bitwise
    const b = dataArray[i + 2] >> 3;
    // const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
    // luminanceData[luminanceIndex] =
    //   0.299 * dataArray[i] +
    //   0.587 * dataArray[i + 1] +
    //   0.114 * dataArray[i + 2];
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray[luminanceIndex] = r + b + g;
  }
  return cachedGrayScaleUintArray;
};

export const clientWaitAsync = (
  gl: WebGL2RenderingContext,
  sync: WebGLSync,
  flags = 0,
  intervalMs = 10,
) => {
  return new Promise<void>((resolve, reject) => {
    const check = () => {
      const res = gl.clientWaitSync(sync, flags, 0);
      if (res === gl.WAIT_FAILED) {
        reject();
        return;
      }
      if (res === gl.TIMEOUT_EXPIRED) {
        setTimeout(check, intervalMs);
        return;
      }
      resolve();
    };
    check();
  });
};

export const readPixelsAsync = (
  gl: WebGL2RenderingContext,
  width: number,
  height: number,
  buffer: Uint8Array,
): Promise<Uint8Array> => {
  const bufpak = gl.createBuffer();
  gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
  gl.bufferData(gl.PIXEL_PACK_BUFFER, buffer.byteLength, gl.STREAM_READ);
  gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
  const sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
  if (!sync) throw new Error('WebGLSync is empty');
  gl.flush();
  return clientWaitAsync(gl, sync, 0, GL_BUFFER_STATUS_CHECK_INTERVAL).then(() => {
    gl.deleteSync(sync);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, buffer);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
    gl.deleteBuffer(bufpak);
    return buffer;
  });
};

export function setAttributes(gl: WebGL2RenderingContext, program: WebGLProgram, name: string) {
  const location = gl.getAttribLocation(program, name);
  gl.enableVertexAttribArray(location);
  gl.vertexAttribPointer(location, 2, gl.FLOAT, false, 0, 0);
}

export function isMobile() {
  const userAgent = navigator.userAgent.toLowerCase();
  return /iphone|ipod|android|ie|blackberry|fennec/.test(userAgent);
}
export const getCameraResolution = () => {
  if (isMobile()) {
    return {
      width: { ideal: document.body.clientWidth > 500 ? 500 : document.body.clientWidth },
      height: { ideal: document.body.clientHeight > 500 ? 500 : document.body.clientHeight },
    };
  }
  return {
    width: { ideal: 1000 },
    height: { ideal: 1000 },
  };
};
