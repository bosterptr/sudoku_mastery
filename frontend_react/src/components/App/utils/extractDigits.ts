import { LayersModel, loadLayersModel, reshape, tensor2d } from '@tensorflow/tfjs';

const indexOfMaxValue = <T>(
  a: T[],
  func: (param: T) => number = ((i) => i) as (param: T) => number,
) => a.reduce((iMax, x, i, arr) => (func(x) > func(arr[iMax]) ? i : iMax), 0);

let model: LayersModel;
const CELLS_IN_ROW = 9;
const CELL_WIDTH = 20;
const TOTAL_CELLS = CELLS_IN_ROW * CELLS_IN_ROW;
const IMAGE_WIDTH = CELLS_IN_ROW * CELL_WIDTH;
const CELL_PIXEL_COUNT = CELL_WIDTH * CELL_WIDTH;

function extractCube(src: Uint8Array, x: number, y: number): Uint8Array {
  if (x < 0 || y < 0 || x + CELL_WIDTH > IMAGE_WIDTH || y + CELL_WIDTH > IMAGE_WIDTH) {
    throw new Error('Starting point out of bounds');
  }
  const cube = new Uint8Array(CELL_WIDTH * CELL_WIDTH);
  for (let row = 0; row < CELL_WIDTH; row += 1) {
    for (let col = 0; col < CELL_WIDTH; col += 1) {
      const srcIndex = (y + row) * IMAGE_WIDTH + (x + col);
      const cubeIndex = row * CELL_WIDTH + col;
      cube[cubeIndex] = src[srcIndex];
    }
  }

  return cube;
}

const init = async () => {
  model = await loadLayersModel('./models/model6.json');
};
const extractDigits = (cells: Uint8Array) => {
  const testDataArray = new Float32Array(IMAGE_WIDTH * IMAGE_WIDTH);
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const x = (i % CELLS_IN_ROW) * CELL_WIDTH;
    const y = Math.floor(i / CELLS_IN_ROW) * CELL_WIDTH;
    const buffer = extractCube(cells, x, y);
    testDataArray.set(buffer, i * CELL_PIXEL_COUNT);
  }
  const testTensor = tensor2d(testDataArray, [TOTAL_CELLS, CELL_PIXEL_COUNT]);
  const reshaped = reshape(testTensor, [TOTAL_CELLS, CELL_WIDTH, CELL_WIDTH, 1]);
  const tensor = model.predict(reshaped);
  if (Array.isArray(tensor)) throw new Error();
  const prediction = tensor.dataSync();
  const predictionData: number[] = [];
  for (const num of prediction) {
    predictionData.push(num);
  }
  let result = '';
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const cellPrediction = predictionData.slice(i * 10, i * 10 + 10);
    const digit = indexOfMaxValue(cellPrediction);
    result += digit === 0 ? '.' : digit;
  }
  return result;
};

export { extractDigits, init };
