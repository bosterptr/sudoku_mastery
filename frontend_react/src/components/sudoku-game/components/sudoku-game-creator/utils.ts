import { ALL_SUDOKU_CELLS, CANVAS_SIZE, SUDOKU_CELLS_IN_ROW } from 'app/core/constants/sudoku';
import { ICellNotes, IMission } from './types';

export const generateNotes = (): ICellNotes[] => {
  const newArray: ICellNotes[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    newArray.push([]);
  }
  return newArray;
};
export const getBlockSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 2 * pixelRatio) / 3;
export const getAdjustedPixelRatio = () => {
  const { devicePixelRatio } = window;
  return devicePixelRatio > 2 ? 2 * devicePixelRatio : 4;
};
export const getCellSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 4 * pixelRatio) / 9;
export const getContextSize = (pixelRatio: number) => CANVAS_SIZE * pixelRatio;
export const getOuterSize = (pixelRatio: number) => 2 * pixelRatio;
export const getThinSize = (pixelRatio: number) => 0.75 * pixelRatio;
export const getX = (index: number) => Math.floor(index / 9);
export const getY = (index: number) => index % 9;
export const getCellXPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getY(index);
export const getCellYPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getX(index);
export const getCubeId = (index: number) =>
  `${Math.floor(getX(index) / 3)}${Math.floor(getY(index) / 3)}`;
export const EMPTY_CELL = '.';
export const isSudokuCellEmpty = (char: string) => char === EMPTY_CELL;
export const generateArrWithTrueAtIndexes = (indexes: number[]): boolean[] => {
  const newArray: boolean[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (indexes.includes(cellIndex)) newArray[cellIndex] = true;
    else newArray[cellIndex] = false;
  }
  return newArray;
};
export const getScoreFromMove = (values: IMission, solution: string, selectedCell: number) => {
  let fullRows = 0;
  let fullColumns = 0;
  let fullCubes = 0;
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (values[cellIndex] === solution[cellIndex]) {
      if (getY(cellIndex) === getY(selectedCell)) fullRows += 1;
      if (getX(cellIndex) === getX(selectedCell)) fullColumns += 1;
      if (getCubeId(cellIndex) === getCubeId(selectedCell)) fullCubes += 1;
    }
  }
  let score = 1;
  if (fullRows === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullColumns === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullCubes === 9) score += 5;
  return score;
};
