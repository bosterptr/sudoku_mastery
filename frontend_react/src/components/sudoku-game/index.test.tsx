/* eslint-disable react/jsx-props-no-spreading */
import { render } from 'app/tests/utils';
import NOOP from 'app/utils/noop';
import { SudokuGame } from '.';
import { ISudoku, SudokuDifficulty } from './types';

const CURRENT_GAME: ISudoku = {
  id: '379',
  body: '3.8.15...4.97....5561..8...9.4...53..5...1.49......8...3...7..67..39.2......8.7..',
  difficulty: SudokuDifficulty.Easy,
  created_at: '',
  updated_at: '',
  user_id: '',
  // winRate: 50,
};

jest.mock('sudoku_wasm', () => {
  return {
    __esModule: true,
    default: jest.fn(),
  };
});
describe('SudokuGame', () => {
  it('renders a game with required props', () => {
    const { asFragment } = render(
      <SudokuGame
        game={CURRENT_GAME}
        onLose={NOOP}
        onWin={NOOP}
        onAcceptedLostGame={NOOP}
        onAcceptedWonGame={NOOP}
        onChangeGameType={NOOP}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
