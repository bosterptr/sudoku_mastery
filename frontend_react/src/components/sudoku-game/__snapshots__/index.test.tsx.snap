// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SudokuGame renders a game with required props 1`] = `
<DocumentFragment>
  .c16 {
  fill: #3d4150;
  stroke: transparent;
}

.c10 {
  position: relative;
  width: 100%;
  flex-basis: 60%;
  min-width: 250px;
  max-width: 500px;
}

.c11 {
  max-width: 100%;
  max-height: 100%;
  cursor: pointer;
}

.c2 {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.c3 {
  font-size: 16px;
  font-weight: 500;
  color: #24292ebf;
}

.c4 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  background-color: #ffffff;
}

.c4:focus {
  outline: none;
  text-decoration: none;
}

.c4:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c4:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c5 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c5:focus {
  outline: none;
  text-decoration: none;
}

.c5:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c5:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c1 {
  align-items: center;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  margin-bottom: 5px;
  position: relative;
}

.c7 {
  font-weight: 400;
  font-size: 16px;
  line-height: 1.43;
  color: #3d4150;
}

.c6 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 38%;
}

.c8 {
  color: #000000;
  font-size: 13px;
  font-weight: 600;
  margin-left: 4px;
}

.c20 {
  background-color: #ffffff;
  border-radius: 5px;
  border: 0;
  color: #1648ff;
  cursor: pointer;
  flex-basis: 32%;
  font-size: 3.2em;
  font-weight: 700;
  height: auto;
  margin: 1%;
  padding: 0px;
  text-align: center;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out;
}

.c20:hover {
  background-color: #f0f4f4;
  color: #1648ff;
}

.c20:active {
  background-color: #1648ff;
  color: #fff;
}

.c19 {
  display: flex;
  flex-wrap: wrap;
  height: 300px;
  width: 300px;
}

.c14 {
  align-items: center;
  background-color: transparent;
  border: 1px solid transparent;
  cursor: pointer;
  display: flex;
  flex-direction: column;
  position: relative;
}

.c15 {
  align-items: center;
  background-color: #ffffff;
  border-radius: 10px;
  border: 1px solid #1648ff;
  color: #1648ff;
  cursor: pointer;
  display: flex;
  font-size: 1.8em;
  justify-content: center;
  min-height: 64px;
  min-width: 64px;
  user-select: none;
  transition: color 0.3s ease,background-color 0.3s ease;
}

.c15:hover {
  color: #1648ff;
  background-color: #f4f5f5;
  fill: #1648ff;
}

.c15:active {
  background-color: #1648ff;
  color: #ffffff;
  fill: #1648ff;
}

.c17 {
  color: #24292ebf;
  font-weight: 600;
  text-align: center;
  white-space: nowrap;
}

.c18 {
  align-items: center;
  background-color: #1648ff;
  border-radius: 50px;
  box-shadow: 0px 1px 2px #181a1b33;
  color: #fff;
  display: flex;
  font-size: 12px;
  font-weight: 700;
  justify-content: center;
  left: 50%;
  padding: 0;
  padding: 8px;
  position: absolute;
  top: -16%;
  user-select: none;
  z-index: 1;
  transition: 0.2s ease background-color,0.2s ease color;
  background-color: #f0f4f4;
  color: #3d4150;
}

.c13 {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  margin-bottom: 10px;
}

.c12 {
  display: flex;
  flex-direction: column;
  margin-left: 20px;
}

.c9 {
  flex-direction: row;
  position: relative;
  display: flex;
  align-items: center;
}

.c0 {
  flex-direction: column;
  display: flex;
  position: relative;
  margin: 0 auto;
}

@media (max-width: 400px) {
  .c20 {
    font-size: 2.4em;
  }
}

@media (min-width: 801px) {
  .c20:nth-child(3n-2) {
    margin-left: 0;
  }

  .c20:nth-child(3n) {
    margin-right: 0;
  }

  .c20:first-child,
  .c20:nth-child(2),
  .c20:nth-child(3) {
    margin-top: 0;
  }

  .c20:nth-child(7),
  .c20:nth-child(8),
  .c20:nth-child(9) {
    margin-bottom: 0;
  }
}

@media (max-width: 800px) {
  .c12 {
    margin-left: 0;
    width: 100%;
  }
}

@media (max-width: 800px) {
  .c9 {
    flex-direction: column;
  }
}

@media (min-width: 768px) {
  .c9 {
    max-width: 768px;
  }
}

@media (min-width: 801px) {
  .c9 {
    min-width: 801px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <div
          class="c3"
        >
          Difficulty:
        </div>
        <div
          class="c4"
        >
          Easy
        </div>
        <div
          class="c5"
        >
          Medium
        </div>
        <div
          class="c5"
        >
          Hard
        </div>
        <div
          class="c5"
        >
          Expert
        </div>
      </span>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          Mistakes:
          <span
            class="c8"
          >
            0/3
          </span>
        </div>
        <div
          class="c7"
        >
          Score:
          <span
            class="c8"
          >
            0
          </span>
        </div>
        <div
          class="c7"
        >
          00:00
        </div>
      </div>
    </div>
    <div
      class="c9"
    >
      <div
        class="c10"
      >
        <div
          class=""
        >
          <canvas
            class="c11"
            data-testid="sudoku-canvas"
            height="2000"
            style="width: 0px; height: 0px;"
            width="2000"
          />
        </div>
      </div>
      <div
        class="c12"
      >
        <div
          class="c13"
        >
          <button
            class="c14"
          >
            <div
              class="c15"
            >
              <svg
                class="c16"
                height="16px"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="16px"
              >
                <path
                  d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
                />
              </svg>
            </div>
            <div
              class="c17"
            >
              Undo
            </div>
          </button>
          <button
            class="c14"
          >
            <div
              class="c15"
            >
              <svg
                class="c16"
                height="16px"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="16px"
              >
                <path
                  d="M497.941 273.941c18.745-18.745 18.745-49.137 0-67.882l-160-160c-18.745-18.745-49.136-18.746-67.883 0l-256 256c-18.745 18.745-18.745 49.137 0 67.882l96 96A48.004 48.004 0 0 0 144 480h356c6.627 0 12-5.373 12-12v-40c0-6.627-5.373-12-12-12H355.883l142.058-142.059zm-302.627-62.627l137.373 137.373L265.373 416H150.628l-80-80 124.686-124.686z"
                />
              </svg>
            </div>
            <div
              class="c17"
            >
              Erase
            </div>
          </button>
          <button
            class="c14"
          >
            <div
              class="c15"
            >
              <svg
                class="c16"
                height="16px"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="16px"
              >
                <path
                  d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                />
              </svg>
            </div>
            <div
              class="c18"
            >
              OFF
            </div>
            <div
              class="c17"
            >
              Notes
            </div>
          </button>
        </div>
        <div
          class="c19 "
        >
          <button
            class="c20"
          >
            1
          </button>
          <button
            class="c20"
          >
            2
          </button>
          <button
            class="c20"
          >
            3
          </button>
          <button
            class="c20"
          >
            4
          </button>
          <button
            class="c20"
          >
            5
          </button>
          <button
            class="c20"
          >
            6
          </button>
          <button
            class="c20"
          >
            7
          </button>
          <button
            class="c20"
          >
            8
          </button>
          <button
            class="c20"
          >
            9
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
